peer/ping_manager.go

```diff
@@ -0,0 +1,262 @@
+	newPingPayload func() []byte
+
+	// newPongSize is a closure that returns a random value between
+	// [0, lnwire.MaxPongBytes]. This random value helps to more effectively
+	// pair Pong messages with Ping.
+	newPongSize func() uint16
+
+	// intervalDuration is the Duration between attempted pings.
+	intervalDuration time.Duration
+
+	// timeoutDuration is the Duration we wait before declaring a ping
+	// attempt failed.
+	timeoutDuration time.Duration
+
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Pointer[time.Duration]
+
+	// pingLastSend is the time when we sent our last ping message.
+	// To be used atomically.
+	pingLastSend *time.Time
+
+	// outstandingPongSize is the current size of the requested pong
+	// payload.  This value can only validly range from [0,65531]. Any
+	// value < 0 is interpreted as if there is no outstanding ping message.
+	outstandingPongSize int32
+
+	// pingTicker is a pointer to a Ticker that fires on every ping
+	// interval.
+	pingTicker *time.Ticker
+
+	// pingTimeout is a Timer that will fire when we want to time out a
+	// ping
+	pingTimeout *time.Timer
+
+	// pingChan is the channel on which the pingManager will write Ping
+	// messages it wishes to send out
+	pingChan chan<- lnwire.Ping
+
+	// pongChan is the channel on which the pingManager will write Pong
+	// messages it is evaluating
+	pongChan <-chan lnwire.Pong
+
+	// failChan is the channel on which the pingManager will report ping
+	// failures. This will happen if the received Pong message deviates
+	// from what is acceptable, or if it times out.
+	failChan chan<- struct{}
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewPingManager constructs a pingManager in a valid state. It must be started
+// before it does anything useful, though.
+// NOTE: Do not manually construct a pingManager, use this function instead.
+func NewPingManager(newPingPayload func() []byte, newPongSize func() uint16,
+	intervalDuration time.Duration, timeoutDuration time.Duration,
+) *PingManager {
+
+	m := PingManager{
+		newPingPayload:      newPingPayload,
+		newPongSize:         newPongSize,
+		intervalDuration:    intervalDuration,
+		timeoutDuration:     timeoutDuration,
+		outstandingPongSize: -1,
+	}
+
+	return &m
+}
+
+// Start launches the primary goroutine that is owned by the pingManager. It
+// returns 3 channels that can be used to communicate with the pingManager.
+// The first is a read-only channel containing Ping messages to be sent out
+// over the wire. The second is a write-only channel that is used to send the
+// Pong responses to the pingManager for evaluation. The final channel is a
+// read-only channel that is used to communicate that the peer has failed the
+// ping request.
+func (m *PingManager) Start(quit chan struct{}) (<-chan lnwire.Ping,
+	chan<- lnwire.Pong, <-chan struct{}, error) {
```

Comment:
On second thought I think I should do the channel construction prior to the Start method. Otherwise we couldn't set up the coroutine graph without already being on the clock for a ping response. I'll move these channel returns to the constructor.